src/builtin.rs

```diff
@@ -0,0 +1,224 @@
+pub fn ps() {
+    match File::open("context:") {
+        Ok(mut file) => {
+            let mut string = String::new();
+            match file.read_to_string(&mut string) {
+                Ok(_) => println!("{}", string),
+                Err(err) => println!("Failed to read: context: {}", err),
+            }
+        }
+        Err(err) => println!("Failed to open file: context: {}", err),
+    }
+}
+
+pub fn pwd() {
+    match env::current_dir() {
+        Ok(path) => {
+            match path.to_str() {
+                Some(path_str) => println!("{}", path_str),
+                None => println!("?"),
+            }
+        }
+        Err(err) => println!("Failed to get current dir: {}", err),
+    }
+}
+
+pub fn read(args: &Vec<String>, variables: &mut BTreeMap<String, String>) {
+    for i in 1..args.len() {
+        if let Some(arg_original) = args.get(i) {
+            let arg = arg_original.trim();
+            print!("{}=", arg);
+            if let Err(message) = stdout().flush() {
+                println!("{}: Failed to flush stdout", message);
+            }
+            if let Some(value_original) = readln() {
+                let value = value_original.trim();
+                set_var(variables, arg, value);
+            }
+        }
+    }
+}
+
+pub fn rm(args: &Vec<String>) {
+    match args.get(1) {
+        Some(path) => {
+            if fs::remove_file(path).is_err() {
+                println!("Failed to remove: {}", path);
+            }
+        }
+        None => println!("No name provided"),
+    }
+}
+
+pub fn rmdir(args: &Vec<String>) {
+    match args.get(1) {
+        Some(path) => {
+            if fs::remove_dir(path).is_err() {
+                println!("Failed to remove: {}", path);
+            }
+        }
+        None => println!("No name provided"),
+    }
+}
+
+pub fn run(args: &Vec<String>, variables: &mut BTreeMap<String, String>) {
+    let path = "/apps/shell/main.bin";
+
+    let mut command = process::Command::new(path);
+    for i in 1..args.len() {
+        if let Some(arg) = args.get(i) {
+            command.arg(arg);
+        }
+    }
+
+    match command.spawn() {
+        Ok(mut child) => {
+            match child.wait() {
+                Ok(status) => {
+                    if let Some(code) = status.code() {
+                        set_var(variables, "?", &format!("{}", code));
+                    } else {
+                        println!("{}: No child exit code", path);
+                    }
+                }
+                Err(err) => println!("{}: Failed to wait: {}", path, err),
+            }
+        }
+        Err(err) => println!("{}: Failed to execute: {}", path, err),
+    }
+}
+
+pub fn sleep(args: &Vec<String>) {
+    let secs = args.get(1).map_or(0, |arg| arg.to_num());
+    thread::sleep_ms(secs as u32 * 1000);
+}
+
+pub fn touch(args: &Vec<String>) {
+    match args.get(1) {
+        Some(file_name) => {
+            if let Err(err) = File::create(file_name) {
```

Comment:
Since this should be moved to core-utils and is not implemented correctly here, I am just going to remove it.